### PR TITLE
libroach: clang-format C++ code.

### PR DESCRIPTION
**Pull request by mberhault**
Release note: None

* run `clang-format` over libroach code.
* slight tweaks to default config
* add `c-deps-fmt` make target
* readme for how to install/use

### CODE REVIEW

**Comment by mberhault** [comment]
Oh, and the reason the config file is in `c-deps/libroach` is that `clang-format` by default searches for `.clang-format` in parent directories.

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/.clang-format, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hcsmb97YMuA6oPFoZ:-L-hcsmb97YMuA6oPFo_:bw8kcym) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/.clang-format#L72)):*
> ```YAML
> PenaltyExcessCharacter: 1000000
> PenaltyReturnTypeOnItsOwnLine: 60
> PointerAlignment: Right
> ```

After working in Go for so long, right-aligned pointers look weird to me. I'd slightly prefer left-aligned pointers, but @petermattis may disagree (and as long as it's automated I don't feel strongly either way). 

---

*[c-deps/libroach/clang.format, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hbbgcIkulFi7gv0zh:-L-hbbgdrW6avCDbp7bp:bwb1dbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/clang.format#L1)):*
> ```
> ---
> ```

Why both `clang.format` and `.clang-format`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-:-L-hdg5GEJfZ_i5YI7G9:bnfp4nl)*

**Comment by petermattis** [comment]
:lgtm: Thanks for doing this!

---

Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/.clang-format, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hcsmb97YMuA6oPFoZ:-L-hfQcGW3I48YwsrR4b:b-dndjpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/.clang-format#L72)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

After working in Go for so long, right-aligned pointers look weird to me. I'd slightly prefer left-aligned pointers, but @petermattis may disagree (and as long as it's automated I don't feel strongly either way). 
</blockquote></details>

I'd prefer left-aligned pointers given how clang-format applies this. Specifically, right-aligned pointers on return values looks really odd to me.

---

*[c-deps/libroach/ccl/db.cc, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hfaFMx4flKiSL1wyg:-L-hfaFMx4flKiSL1wyh:b-a2q43i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/ccl/db.cc#L9)):*
> ```cpp
> //     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
> 
> #include "../db.h"
> ```

That's weird. Is there a way to make local includes sorted after system includes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-:-L-hfijDTQ7hdSUmFEEe:b-pt8m11)*

**Comment by mberhault** [comment]
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/.clang-format, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hcsmb97YMuA6oPFoZ:-L-hhlMC0rnFE3EZZtMs:b-grhiwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/.clang-format#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'd prefer left-aligned pointers given how clang-format applies this. Specifically, right-aligned pointers on return values looks really odd to me.
</blockquote></details>

It can definitely do left-aligned pointers. I agree that right-aligned pointers also look weird, but it's technically correct when you think about multi-pointer declarations on the same line. The pointer is attached to the variable, not the type.
Still, I'm with Ben, I don't care as long as it does it for me, so left-aligned it it.

---

*[c-deps/libroach/clang.format, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hbbgcIkulFi7gv0zh:-L-hidQt-nhUDG1Q6Tip:b-k73dre) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/clang.format#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why both `clang.format` and `.clang-format`?
</blockquote></details>

erroneous addition. `clang.format` was where I kept for dumped config for comparison. Removed.

---

*[c-deps/libroach/ccl/db.cc, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hfaFMx4flKiSL1wyg:-L-hjV_NY_van1Lr-ChK:b-tjd23s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/ccl/db.cc#L9)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That's weird. Is there a way to make local includes sorted after system includes?
</blockquote></details>

We can define include priorities using regexps. Be aware though that the main header (the file's corresponding .h) is always set to the beginning.
Would something like this be ok:
```
"myfile.h"
<angle includes>
"quoted includes"
```

Given that they're regexps, we could also categorize `<.*/.*>` but I'm not sure that's important.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535)*

**Review by a-robinson — APPROVED** [review]
Thanks for doing this!

File: c-deps/libroach/.clang-format [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+CommentPragmas:  '^ IWYU pragma:'
+ConstructorInitializerAllOnOneLineOrOnePerLine: true
+ConstructorInitializerIndentWidth: 4
+ContinuationIndentWidth: 4
+Cpp11BracedListStyle: true
+DerivePointerAlignment: false
+DisableFormat:   false
+ExperimentalAutoDetectBinPacking: false
+ForEachMacros:   [ foreach, Q_FOREACH, BOOST_FOREACH ]
+IncludeCategories:
```
Comment (mberhault) [inline c-deps/libroach/.clang-format]: @petermattis: this is the include regexp.
Includes are sorted by ascending priority.
Priorities 2 and 3 are the built-in ones.
The current file's include (eg: `db.h` for `db.cc`) remains priority 0, but we can change that regexp too.

File: c-deps/libroach/db.cc [hunk 2]
```diff
@@ -12,9 +12,10 @@
 // implied.  See the License for the specific language governing
 // permissions and limitations under the License.
 
+#include "db.h"
```
Comment (mberhault) [inline c-deps/libroach/db.cc]: @petermattis: how's this?

**Comment by tschottdorf** [comment]
Looks good! Can you make sure to include instructions on running `clang-format` and mentioning it as a requirement (at least when C++ is touched)? Perhaps a `README.md` in `c-deps/libroach{,ccl}` plus a link to it from `CONTRIBUTING.md` (and while you're there, add a similar link to `pkg/ui/README.md` for ui development).

**Comment by mberhault** [comment]
The main thing left that bothers me is the single-line constructor initializers.
There is an option in a later `clang-format` version to always break, but not all.

This is actually a problem that will come up. If people are running different versions, we'll be having formatting conflicts constantly. They're apparently not quite as good as `gofmt` when it comes to modifications over time.

**Comment by petermattis** [comment]
:shipit: 

---

Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/db.cc, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hVZjIfSZn14xqDvs--r3-15:-L-hp6M1Gf1rki6Lpinm:bnhoihu) ([raw file](https://github.com/cockroachdb/cockroach/blob/62c37d850970843f98bd593663580dd24cacb40a/c-deps/libroach/db.cc#L15)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

@petermattis: how's this?
</blockquote></details>

Weird, but I'll get used to it. I think this became the Google standard at some point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-:-L-hpROpOswhU0vFIOGg:bizlbj6)*

**Comment by mberhault** [comment]
added simple readme and `make c-deps-fmt` target to format all non-proto libroach source files.
I'm hesitant to add a lint check quite yet, at least not until I've done some more research on version discrepancies.

File: Makefile [hunk 3]
```diff
@@ -1116,6 +1116,15 @@ $(SQLPARSER_ROOT)/help_messages.go: $(SQLPARSER_ROOT)/sql.y $(SQLPARSER_ROOT)/he
 	mv -f $@.tmp $@
 	gofmt -s -w $@
 
+CLANG_FORMAT := $(shell which clang-format)
```
Comment (mberhault) [inline Makefile]: @benesch: my make-fu is rather weak. Could you take a quick look at this new rule?

**Comment by mberhault** [comment]
PTAL. I think this is good for now. Note the llvm 5.0 restriction.

File: Makefile (hunk 3 above)
Comment (benesch) [reply Makefile]: Looks good! (We don't usually bother to check whether tools are installed, FWIW, and just live with the `/bin/bash: BLAH: command not found` error.)

**Review by benesch — APPROVED** [review]

**Comment by mberhault** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[Makefile, line 1119 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hvSu57nMKpJT_eFaT-r6-1119:-L-ia_CkGO93B4_poZjh:btacaha) ([raw file](https://github.com/cockroachdb/cockroach/blob/228178f5e8e4120d1b50626f69daf51ca86d8fe8/Makefile#L1119)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Looks good! (We don't usually bother to check whether tools are installed, FWIW, and just live with the `/bin/bash: BLAH: command not found` error.)
</blockquote></details>

ok, switched to a plain call.

---

*[c-deps/libroach/.clang-format, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535#-L-hVZjJNMH7Th5_QFFF-r2-72:-L-iU47Bkp6n9ksA93_7:bsl6mk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7bb5b6d932eb8afd1a47fd225038f16e928912/c-deps/libroach/.clang-format#L72)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I'll cast another vote in favor of `PointerAlignment: Left`.
</blockquote></details>

yeah, that's all been converted already

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20535)*

**Pull request merged**